### PR TITLE
JSON/GUI tool editor

### PR DESCRIPTION
**Pull request by matebordas**
I added some components to switch between the GUI and the JSON editing of the tools. There is not much logic added to the components, I mainly focused on the components UI behaviour and making them modular. 
Things that still need to be done: add logic for the CWL files, there are some smaller components (e.g. the input component could be integrated but it needs to be re-factored), also the file directory placement could be better (not sure how to approach this, yet).

### CODE REVIEW

File: client/src/app/components/forms/inputs/forms/docker-input-form.component.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    directives: [DockerInput, InputFromComponent],
+    template: `
+            <input-form [primaryLabel]="'Docker image'" 
+                        [secondaryLabel]="'Docker Repository'"
+                        [inputData]="inputData"
+                        [contentComponent]="contentComponent">
+            </input-form>
+    `
+})
+export class DockerInputForm implements OnInit {
```
Comment (mayacoda) [inline client/src/app/components/forms/inputs/forms/docker-input-form.component.ts]: We should be consistent about naming our classes. From what I've seen in the Angular tutorials, the practice is to include the class type in the class name. This class has a `@Component` decorator, so it should be called `DockerInputFormComponent`.

Granted, it is a bit wordy, but I think the intent is to prevent confusion when importing the class into a different file. If it has Component in the name, its purpose is obvious.

It's up for discussion, though, certainly.

**Comment by ivanbatic** [comment]
Relates to multiple files:
- Use access modifiers
- Remove empty methods
- When declaring properties, put a newline after a decorator for consistency and readability  ( for example, `@assignable @Input() public myAttribute: Subject<FileModel> = new Subject<FileModel>();` is not really nice).
- Use `<button>` instead of `<a>` for action controls. The `<a>` tag represents a link to another document or another part of the same document. If you want a button to look like a link, use CSS.
- Avoid using the `<any>` type

File: client/src/app/components/forms/inputs/types/docker-input.component.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    selector: 'docker-input',
+    template: `
+         <input name="dockerPull"
+                type="text"
+                class="form-control"
+                id="dockerImage"
+                [(ngModel)]="dockerPull">
+    `,
+})
+export class DockerInput implements OnInit {
```
Comment (ivanbatic) [inline client/src/app/components/forms/inputs/types/docker-input.component.ts]: Why is this needed as a separate component?

Comment (matebordas) [inline client/src/app/components/forms/inputs/types/docker-input.component.ts]: It doesn't have to be. My idea was that if we have an input component that is generic enough to attach different validation, we could use that and pass it the methods we want to execute on these occasions.

Comment (mayacoda) [inline client/src/app/components/forms/inputs/types/docker-input.component.ts]: Docker is something I'd really only see being used in one place: in the docker section of a command line tool. So long as we have a form, its inputs don't have to be isolated as well.

File: client/src/app/components/tool-container/tool-header/tool-header.component.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+})
+export class ToolHeader implements OnInit {
+    /** The current view mode is needed for styling the selected button */
+    @Input() viewMode: ViewMode;
+    
+    /** Emit changes of the view mode */
+    @Output() viewModeChanged = new EventEmitter();
+    
+    /** Tool revisions. TODO: load actual revisions */
+    revisions: Array<string> = ["rev1", "rev2", "rev3"];
```
Comment (mayacoda) [inline client/src/app/components/tool-container/tool-header/tool-header.component.ts]: This can wait till we hook up the SB API. For local development, there won't be explicit revisions.

Comment (ivanbatic) [inline client/src/app/components/tool-container/tool-header/tool-header.component.ts]: @mayacoda  Is there maybe a chance we could check if the local project uses VCS and use that for revisions?
For example, list the `git log --follow myTool.yaml` output?

Comment (mayacoda) [inline client/src/app/components/tool-container/tool-header/tool-header.component.ts]: @ivanbatic We certainly can, but that's waaaaay out of scope right now. Feel free to add it to the backlog :)

File: client/src/app/components/tool-container/tool-header/tool-header.component.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+@Component({
+    selector: "tool-header",
+    directives: [NgSelectOption],
+    template: `
+        <header>
+            <select id="revisionSelect" name="rev" [ngModel]="selectedRevision" (change)="onRevisionChange($event)">
+                <option *ngFor="let revision of revisions" [value]="revision">{{revision}}</option>
+            </select>
+                            <span id="guiJsonButtons">
```
Comment (mayacoda) [inline client/src/app/components/tool-container/tool-header/tool-header.component.ts]: Avoid using ids for components that are reusable. According to [the MDN](https://developer.mozilla.org/en/docs/Web/HTML/Global_attributes/id): _the id global attribute defines a unique identifier (ID) which must be unique in the whole document_. It's impossible to ensure that an id in a component will remain unique across the whole document.

File: client/src/app/components/tool-container/tool-header/tool-header.component.scss [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        color: #000000;
+        margin-right: 40px;
+        width: 100px;
+    }
+
+    button {
+        width: 8%;
+    }
+
+    #guiJsonButtons {
```
Comment (mayacoda) [inline client/src/app/components/tool-container/tool-header/tool-header.component.scss]: Likewise, avoid using ids as CSS selectors. Because they are unique, any [qualifier before an id is superfluous](https://developer.mozilla.org/en-US/docs/Web/Guide/CSS/Writing_efficient_CSS#Guidelines_for_Efficient_CSS) and makes matching selector to element slower. Though now that I think about it, it would be pretty useful to apply everything in this guide to all our CSS. We don't have much of it, but it will grow.

File: client/src/app/components/tool-container/tool-container.component.scss [hunk 6]
```diff
@@ -0,0 +1,21 @@
+@import "~bootstrap/scss/mixins";
+@import "~bootstrap/scss/_variables";
+@import "~bootstrap/scss/_forms";
+@import "~bootstrap/scss/_input-group";
+@import "~bootstrap/scss/_tooltip";
+@import "~bootstrap/scss/_buttons";
+@import "~bootstrap/scss/_button-group";
+@import "~bootstrap/scss/_close";
```
Comment (mayacoda) [inline client/src/app/components/tool-container/tool-container.component.scss]: For reference, you can (should) just uncomment these imports in [_bootstrap-pack.scss ](https://github.com/rabix/cottontail/blob/next/client/src/assets/sass/_bootstrap-pack.scss)

File: client/src/app/components/tool-container/tool-header/tool-header.component.scss (hunk 5 above)
Comment (ivanbatic) [inline client/src/app/components/tool-container/tool-header/tool-header.component.scss]: Also, `header` is a standard DOM tag for which we shouldn't assume a particular styling.

File: client/src/app/components/gui-editor/commandline/commandline.component.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+} from "@angular/core";
+import {NgFor} from "@angular/common";
+import {VisibilityState} from "../animation.states";
+
+require ("./commandline.component.scss");
+
+/** TODO: make this switch between an expression editor and an object inspector*/
+@Component({
+    selector: "commandline",
+    directives: [NgFor],
```
Comment (mayacoda) [inline client/src/app/components/gui-editor/commandline/commandline.component.ts]: This is an unnecessary import.

File: client/src/app/components/gui-editor/gui-editor.component.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+import {EditorSidebarComponent} from "./sidebar/editor-sidebar.component";
+import {PropertyPosition, VisibilityState} from "./animation.states";
+import {CommandLineComponent} from "./commandline/commandline.component";
+
+require("./gui-editor.component.scss");
+
+@Component({
+    selector: "gui-editor",
+    providers: [GuiEditorService],
+    directives: [NgFor, PropertyInputComponent, EditorSidebarComponent, CommandLineComponent],
```
Comment (mayacoda) [inline client/src/app/components/gui-editor/gui-editor.component.ts]: You don't need to import core directives like `NgFor` anymore: http://stackoverflow.com/questions/34385566/how-to-import-ngfor-in-angular-2-using-javascript#answer-34391950

File: client/src/app/components/gui-editor/gui-editor.events.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import {SidebarType} from "./sidebar/editor-sidebar.component";
+
+export enum GuiEditorEventType {
+    showSidebar
```
Comment (mayacoda) [inline client/src/app/components/gui-editor/gui-editor.events.ts]: Convention for enum values is [PascalCase](https://www.typescriptlang.org/docs/handbook/enums.html), I believe.

File: client/src/app/components/gui-editor/sidebar/editor-sidebar.component.scss [hunk 10]
```diff
@@ -0,0 +1,23 @@
+@import "../../../../assets/sass/variables";
+
+.rightSidebar {
```
Comment (ivanbatic) [inline client/src/app/components/gui-editor/sidebar/editor-sidebar.component.scss]: kebab-case for css class names

File: client/src/app/components/gui-editor/sidebar/editor-sidebar.component.ts [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    state,
+    transition,
+    trigger,
+    EventEmitter,
+    Output
+} from "@angular/core";
+import {NgFor} from "@angular/common";
+import {GuiEditorService} from "../gui-editor.service";
+import {VisibilityState} from "../animation.states";
+import {GuiEditorEvent, GuiEditorEventType, ShowSidebarEvent} from "../gui-editor.events";
```
Comment (mayacoda) [inline client/src/app/components/gui-editor/sidebar/editor-sidebar.component.ts]: Remove unused imports. You can do `Optimize imports` in IntelliJ Idea by pressing  `⌃` + `⌥` + `O`, it'll remove imports that aren't being used.

EDIT: it will not remove imports that aren't valid, in this case `GuiEditorEvent` which does not exist in `../gui-editor.events`.

File: client/src/app/components/gui-editor/sidebar/editor-sidebar.component.ts [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+                            <!-- TODO: add expression and objects editor -->
+                            This is the right sidebar content
+                        </div>
+                    </div>
+    `
+})
+export class EditorSidebarComponent {
+    /** Emit changes of the sidebar animation to the parent component */
+    @Output()
+    private sidebarVisibility = new EventEmitter();
```
Comment (ivanbatic) [inline client/src/app/components/gui-editor/sidebar/editor-sidebar.component.ts]: Why an EventEmmiter instead of an observable subject?
Information for `sidebarVisibility` and `sidebarState` can be contained in the BehaviorSubject.

File: client/src/app/components/gui-editor/sidebar/editor-sidebar.component.ts [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+
+    /** State of the sidebar animation */
+    private sidebarState: VisibilityState = "hidden";
+
+    constructor(private guiEditorService: GuiEditorService) {
+        const self = this;
+
+        this.guiEditorService.publishedSidebarEvents.subscribe((event: ShowSidebarEvent) => {
+            self.showSideBar(event.data.sidebarType);
```
Comment (ivanbatic) [inline client/src/app/components/gui-editor/sidebar/editor-sidebar.component.ts]: `Self` is not needed.
Arrow functions bind the parent context (class instance `this`)

File: client/src/app/components/gui-editor/gui-editor.component.ts [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+            state("center", style({
+                margin: '20px auto'
+            })),
+            transition("hidden => visible", animate("100ms ease-in")),
+            transition("visible => hidden", animate("100ms ease-out"))
+        ])
+    ],
+    template: `
+            <main>
+                <property-input @propertyPosition="propertyPosition"
```
Comment (ivanbatic) [inline client/src/app/components/gui-editor/gui-editor.component.ts]: Why are you doing the whole property-input this way?
Why not make a standard form with standard form inputs?

File: client/src/app/components/tool-container/tool-header/tool-header.component.scss [hunk 15]
```diff
@@ -0,0 +1,19 @@
+@import "../../../../assets/sass/variables";
+
+.toolHeaderBtn {
+    width: 8%;
```
Comment (mayacoda) [inline client/src/app/components/tool-container/tool-header/tool-header.component.scss]: Avoid setting width and height on buttons, because it can lead to situations where the text no longer fits inside the button. If you set a fixed width/height in pixels, text will overflow if the user increases the font-size. If you set a percent width/height, at a certain point the text will also overflow because the button is too small. It's better to either use [bootstrap sizes](http://v4-alpha.getbootstrap.com/components/buttons/#sizes) like `btn-sm`, `btn-lg` etc, or to set a custom padding on the button to make it smaller or larger.

File: client/src/app/components/tool-container/tool-header/tool-header.component.ts [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+    directives: [NgSelectOption],
+    template: `
+        <header class="toolHeader">
+                <span class="guiJsonButtons">
+                    <button type="button"
+                            class="btn btn-secondary btn-sm selected toolHeaderBtn"
+                            [ngClass]="{selected: viewMode === 'json'}"
+                            (click)="changeViewMode('json')">JSON</button>
+
+                    <button type="button toolHeaderBtn"
```
Comment (mayacoda) [inline client/src/app/components/tool-container/tool-header/tool-header.component.ts]: `toolHeaderBtn` is not a valid value for the `type` attribute of a  `<button>` element.

File: client/src/app/components/gui-editor/gui-editor.component.ts (hunk 14 above)
Comment (matebordas) [inline client/src/app/components/gui-editor/gui-editor.component.ts]: I will change it. I did it this way because I thought our UI depends on the JSON model we get.

File: client/src/app/components/gui-editor/sidebar/editor-sidebar.component.ts (hunk 12 above)
Comment (matebordas) [inline client/src/app/components/gui-editor/sidebar/editor-sidebar.component.ts]: Seemed simpler to listen to an event on the parent, instead of sending it through a service. I'm using this as directive, so I would not like to access the class property directly. (If that is what you meant)

File: client/src/app/components/tool-container/tool-header/tool-header.component.scss (hunk 15 above)
Comment (matebordas) [inline client/src/app/components/tool-container/tool-header/tool-header.component.scss]: I rather not use Bootstrap sizes, if we decide to get rid of bootstrap.

File: client/src/app/components/gui-editor/sidebar/editor-sidebar.component.ts (hunk 12 above)
Comment (ivanbatic) [inline client/src/app/components/gui-editor/sidebar/editor-sidebar.component.ts]: You don't need a service in order to use observables and subjects. Just by using a BehaviorSubject, you wouldn't need to externally track the current state of the sidebar inside this class, and you would be able to get the current state outside of it, wherever your are listening to the `sidebarVisibility`. No need to access class properties.

Comment (matebordas) [inline client/src/app/components/gui-editor/sidebar/editor-sidebar.component.ts]: something like?

```
//parent
...
temlpate:` 
    <sidebar visiblity="sidebarVisibility">
`
...
class ParentComponent  {
   sidebarVisibility = new BehaviourSubject();

   constructor() {
       ....
      this.sidebarVisibility.subscribe( //do something);
  }
}

//sidebar
visiblity.next("visible")
```

EDIT: because the EventEmitter is actually a Subject so this is technically the same.

**Comment by matebordas** [comment]
I will create a new branch from /next and merge the the /gui_editor there to resolve the conflicts, then send a new pull request.

**Comment by ivanbatic** [comment]
You can just merge the upstream changes from `cottontail/next` into `matesk8:gui_editor` locally, and push the changes onto this branch.

File: client/src/app/components/tool-container/tool-header/tool-header.component.ts [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+    Output
+} from "@angular/core";
+import {NgSelectOption} from "@angular/common";
+import {ViewMode} from "../tool-container.component";
+
+require("./tool-header.component.scss");
+
+@Component({
+    selector: "tool-header",
+    directives: [NgSelectOption],
```
Comment (ivanbatic) [inline client/src/app/components/tool-container/tool-header/tool-header.component.ts:14]: This seems like an unneeded import?
